examples/linear_deformation.py

```diff
@@ -0,0 +1,1187 @@
+
+        def_grad = template_grad.space.element([gf * func for gf in def_grad])
+
+        return def_grad
+
+
+class ShapeRegularizationFunctional(Operator):
+
+    """Regularization functional for linear shape deformations.
+
+    The shape regularization functional is given as
+
+        S(alpha) = 1/2 * ||v(alpha)||^2 = 1/2 * alpha^T K alpha,
+
+    where ``||.||`` is the norm in a reproducing kernel Hilbert space
+    given by parameters ``alpha``. ``K`` is the kernel matrix.
+    """
+    # TODO: let user specify K
+
+    def __init__(self, par_space, ft_kernel):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        par_space : `ProductSpace`
+            Parameter space of the deformations, i.e. the space of the
+            ``alpha_k`` parametrizing the deformations
+        kernel_op : `numpy.ndarray` or `Operator`
+            The kernel matrix defining the norm. If an operator is
+            given, it is called to evaluate the matrix-vector product
+            of the kernel matrix with a given ``alpha``.
+        """
+        super().__init__(par_space, odl.RealNumbers(), linear=False)
+#        if isinstance(kernel_op, Operator):
+#            self._kernel_op = kernel_op
+#        else:
+#            self._kernel_op = odl.MatVecOperator(kernel_op)
+        self.par_space = par_space
+        self.ft_kernel = ft_kernel
+
+    def _call(self, alphas):
+        """Return ``self(alphas)``."""
+        # TODO: how to improve
+
+        # Compute the shape energy by fft
+        ft_momenta = vectorial_ft_shape_op(alphas)
+        stack = vectorial_ft_shape_op.inverse(self.ft_kernel * ft_momenta)
+        return sum(s.inner(s.space.element(
+                       np.asarray(a).reshape(-1, order=self.domain[0].order)))
+                   for s, a in zip(stack, alphas)) / 2
+
+#        # Compute the shape energy by matrix times vector
+#        stack = [self._kernel_op(
+#                     np.asarray(a).reshape(-1, order=self.domain[0].order))
+#                 for a in alphas]
+#        return sum(s.inner(s.space.element(
+#                       np.asarray(a).reshape(-1, order=self.domain[0].order)))
+#                   for s, a in zip(stack, alphas)) / 2
+
+    def _gradient(self, alphas):
+        """Return the gradient at ``alphas``.
+
+        The gradient of the functional is given by
+
+            grad(S)(alpha) = K alpha
+        """
+#        return self.domain.element([self._kernel_op(np.asarray(a).reshape(-1))
+#                                    for a in alphas])
+        pass
+
+    def _gradient_ft(self, alphas):
+        """Return the gradient at ``alphas``.
+
+        The gradient of the functional is given by
+
+            grad(S)(alpha) = K alpha.
+
+        This is used for the n-D case: control grid = image grid.
+        """
+        ft_momenta = vectorial_ft_shape_op(alphas)
+        return vectorial_ft_shape_op.inverse(self.ft_kernel * ft_momenta)
```

Comment:
Same as above: `vectorial_ft_shape_op` is specific to your example and will not be defined if the example code is gone.